### PR TITLE
Pic 2167

### PR DESCRIPTION
**Pull request by vb-moj**

### CODE REVIEW

**Comment by vb-moj** [comment]
Working on the tests  ATM.

File: src/main/java/uk/gov/justice/probation/courtcaseservice/controller/OffenderMatchesController.java [hunk 1]
```diff
@@ -48,28 +53,38 @@ Mono<ResponseEntity<Object>> createGroupedOffenderMatchesByDefendant(@PathVariab
     @GetMapping(value = "/case/{caseId}/defendant/{defendantId}/grouped-offender-matches/{groupId}", consumes = APPLICATION_JSON_VALUE, produces = APPLICATION_JSON_VALUE)
     @ResponseStatus(HttpStatus.OK)
     public @ResponseBody
-    Mono<GroupedOffenderMatchesEntity> getOffenderMatchesByCaseId(@PathVariable(value = "caseId") String caseId,
-                                                                  @PathVariable(value = "defendantId") String defendantId,
-                                                                  @PathVariable(value = "groupId") Long groupId) {
+    Mono<GroupedOffenderMatchesEntity> getOffenderMatchesByCaseId(@PathVariable(value = CASE_ID_PARAM) String caseId,
```
Comment (jhevans) [inline src/main/java/uk/gov/justice/probation/courtcaseservice/controller/OffenderMatchesController.java:56]: I'm not sure why we've specified the `@PathVariable` values here at all, they're redundant as the parameter name already matches what's in the URL so I'd get rid of them if we can.

File: src/main/java/uk/gov/justice/probation/courtcaseservice/controller/OffenderMatchesController.java [hunk 2]
```diff
@@ -48,28 +53,38 @@ Mono<ResponseEntity<Object>> createGroupedOffenderMatchesByDefendant(@PathVariab
...
-    Mono<GroupedOffenderMatchesEntity> getGroupedOffenderMatchesEntityByDefendantIdAndGroupId(@PathVariable(value = "defendantId") String defendantId,
-                                                                                              @PathVariable(value = "groupId") Long groupId) {
-        return offenderMatchService.getGroupedOffenderMatchesEntityByDefendantIdAndGroupId(defendantId, groupId);
+    Mono<GroupedOffenderMatchesEntity> getGroupedOffenderMatchesByDefendantIdAndGroupId(@PathVariable(value = DEFENDANT_ID_PARAM) String defendantId,
+                                                                                        @PathVariable(value = GROUP_ID_PARAM) Long groupId) {
+        return offenderMatchService.getGroupedOffenderMatchesByDefendantIdAndGroupId(defendantId, groupId);
+    }
+
+    @Operation(description = "Creates a new grouped-offender-match entity associated with a defendant")
+    @PostMapping(value = CREATE_GROUPED_OFFENDER_MATCHES_ENTITY_FOR_DEFENDANT_ID_PATH, consumes = APPLICATION_JSON_VALUE, produces = APPLICATION_JSON_VALUE)
```
Comment (jhevans) [inline src/main/java/uk/gov/justice/probation/courtcaseservice/controller/OffenderMatchesController.java:81]: Is there a reason you've extracted the url template as a constant here?

File: src/main/java/uk/gov/justice/probation/courtcaseservice/service/OffenderMatchService.java [hunk 3]
```diff
@@ -129,13 +134,32 @@ Sentence getSentenceForMostRecentConviction(List<Conviction> convictions) {
...
+                    if (hearingEntity != null) {
+                        return createOrUpdateGroupedMatchesByDefendant(hearingEntity, defendantId, groupedOffenderMatchesRequest);
+                    }
+                    throw new EntityNotFoundException(String.format("Hearing  entity not found for defendant %s", defendantId));
+                })
+                .map(groupedOffenderMatchRepository::save);
+
+    }
+
+    private GroupedOffenderMatchesEntity createOrUpdateGroupedMatchesByDefendant(HearingEntity hearingEntity, String defendantId, GroupedOffenderMatchesRequest offenderMatches) {
```
[the remark this answers is not in the document]
Comment (vb-moj) [reply src/main/java/uk/gov/justice/probation/courtcaseservice/service/OffenderMatchService.java:160]: Yes, but there is already one that exists with a different approach.

File: src/main/java/uk/gov/justice/probation/courtcaseservice/controller/OffenderMatchesController.java (hunk 2 above)
Comment (vb-moj) [reply src/main/java/uk/gov/justice/probation/courtcaseservice/controller/OffenderMatchesController.java:81]: Just for better readability.

File: src/main/java/uk/gov/justice/probation/courtcaseservice/controller/OffenderMatchesController.java (hunk 1 above)
Comment (vb-moj) [reply src/main/java/uk/gov/justice/probation/courtcaseservice/controller/OffenderMatchesController.java:56]: Agree, will remove them all.

**Review by rschallapalli — COMMENTED** [review]
couple of minor suggestions

File: src/main/java/uk/gov/justice/probation/courtcaseservice/controller/OffenderMatchesController.java [hunk 4]
```diff
@@ -5,13 +5,7 @@
...
 import org.springframework.http.HttpStatus;
 import org.springframework.http.ResponseEntity;
-import org.springframework.web.bind.annotation.GetMapping;
-import org.springframework.web.bind.annotation.PathVariable;
-import org.springframework.web.bind.annotation.PostMapping;
-import org.springframework.web.bind.annotation.RequestBody;
-import org.springframework.web.bind.annotation.ResponseBody;
-import org.springframework.web.bind.annotation.ResponseStatus;
-import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.bind.annotation.*;
```
[the remark this answers is not in the document]
Comment (vb-moj) [reply src/main/java/uk/gov/justice/probation/courtcaseservice/controller/OffenderMatchesController.java:8]: IntelliJ did this. easy to remove anyway.

**Review by jhevans — APPROVED** [review]

**Review by rschallapalli — APPROVED** [review]

**Pull request merged**